src/pydvl/utils/parallel/futures.py

```diff
@@ -0,0 +1,81 @@
+from concurrent.futures import Executor, Future
+from contextlib import contextmanager
+from dataclasses import asdict
+from typing import Callable, Generator, TypeVar
+from weakref import WeakSet
+
+import ray
+
+from ..config import ParallelConfig
+from .backend import effective_n_jobs
+
+__all__ = ["init_executor", "RayExecutor"]
+
+T = TypeVar("T")
+
+
+@contextmanager
+def init_executor(
+    max_workers: int, config: ParallelConfig
+) -> Generator[Executor, None, None]:
+    """Initializes a futures executor based on the passed parallel configuration object.
+
+    :param max_workers: Maximum number of concurrent tasks.
+    :param config: instance of :class:`~pydvl.utils.config.ParallelConfig` with cluster address, number of cpus, etc.
+
+    :Example:
+
+    >>> from pydvl.utils.parallel.futures import init_executor
+    >>> from pydvl.utils.config import ParallelConfig
+    >>> config = ParallelConfig(backend="ray")
+    >>> with init_executor(max_workers=4, config=config) as executor:
+    ...     pass
+
+    """
+    if config.backend == "ray":
+        max_workers = effective_n_jobs(max_workers, config=config)
+        executor = RayExecutor(max_workers, config)
+    else:
+        raise NotImplementedError(f"Unexpected parallel type {config.backend}")
+    yield executor
+
+
+class RayExecutor(Executor):
```

Comment:
I will a few things taking into account ideas from the linked PR. In the current implementation we get `map()` for free by inhereting from the base `Executor` class just like it is done for `ThreadPoolExecutor` and `ProcessPoolExecutor`.

Reply:
Note that the number of chunks is a dummy in the base's map. Only the process executor chunkifies.

Reply:
Since we rely on ray and don't implement our own workers we cannot control chunking without modifying the function that we're submitting (we could wrap it in another function and submit that instead though but for I think it would add too much complexity).

I added more code to throttle the number of submitted futures.